c2_utils/src/mfx_gralloc4.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2017-2021 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "mfx_gralloc4.h"
+#include "cros_gralloc/cros_gralloc_helpers.h"
+#include "cros_gralloc/i915_private_android.h"
+#include "mfx_c2_utils.h"
+#include "mfx_debug.h"
+#include "mfx_c2_debug.h"
+
+#undef MFX_DEBUG_MODULE_NAME
+#define MFX_DEBUG_MODULE_NAME "mfx_gralloc4"
+
+using aidl::android::hardware::graphics::common::PlaneLayout;
+
+c2_status_t MfxGralloc4Module::Init()
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    sp<IMapper4> mapper = IMapper4::getService();
+    if (nullptr == mapper)
+    {
+        return C2_CORRUPTED;
+    }
+
+    m_mapper = mapper;
+    
+    return C2_OK;
+}
+
+MfxGralloc4Module::~MfxGralloc4Module()
+{
+
+}
+
+Error4 MfxGralloc4Module::get(const native_handle_t* handle, const IMapper4::MetadataType& metadataType,
+                    hidl_vec<uint8_t>& outVec)
+{
+    Error4 err;
+    if (nullptr == m_mapper)
+        return Error4::NO_RESOURCES;
+    m_mapper->get(const_cast<native_handle_t*>(handle), metadataType,
+                [&](const auto& tmpError, const hidl_vec<uint8_t>& tmpVec)
+                {
+                    err = tmpError;
+                    outVec = tmpVec;
+                });
+    return err;
+}
+
+c2_status_t MfxGralloc4Module::GetBufferDetails(const buffer_handle_t handle, BufferDetails *details)
+{
+    MFX_DEBUG_TRACE_FUNC;
+    c2_status_t res = C2_OK;
+
+    do
+    {
+        details->handle = handle;
+
+        auto cros_hnd = cros_gralloc_convert_handle(handle);
+        details->prime = cros_hnd->fds[0];
+        MFX_DEBUG_TRACE_I32(details->prime);
```

Comment:
To get prime fd, maybe need add a new api in minigbm, similar to  "GRALLOC1_FUNCTION_GET_PRIME" in i915_private_android_types.h, like define a "android::gralloc4::MetadataType_PrimeFd"? @feijiang1 @yhe39 @renchenglei